File: DS _LAB_SEM_3/Unit-2 Data Structures_ Arrays,Stack,Queue and Linked List/Lab_11/Cir_Linked_List.java
import java.util.Scanner;

class LinkedList
{
    static Node head;
    static Node last;

    static class Node
    {
        int data;
        Node next;

        public Node(int data)
        {
            this.data = data;
            this.next = null;
        }
    }

    public void addFront(int data)
    {
        Node newNode = new Node(data);

        if(head==null)
        {
            newNode.next = newNode;
            head = newNode;
            last = newNode;
        }
        else
        {
            newNode.next = head;
            last.next = newNode;
            head = newNode;
        }
    }

    public void insertEnd(int data)
    {
        Node newNode = new Node(data);

        if (head == null) 
        {
            newNode.next = newNode;
            head = newNode;
            last = newNode;
        } 
        else 
        {
            newNode.next = head;
            last.next = newNode;
            last = newNode;
        }
    }

    public void display()
    {
        if(head==null)
        {
            System.out.println("Empty");
            return;
        }

        Node save = head; 
        
        do
        {
            System.out.print(save.data + " ");
            save = save.next;
        }while (save != head);

        System.out.println();
    }

    public void deleteAtPosition(int position) 
    {
        if (head == null) 
        {
            System.out.println("Empty");
            return;
        }

        // IF AND OLNY IF ONE NODE IN LINKED LIST THEN FLOW BELOW CODE
        if (position == 1) 
        {
            if (head == last) 
            {
                head = null;
                last = null;
            } 
            else 
            {
                Node deleteNode = head;
                last.next = head.next;
                head = head.next;
            }
        return;
        }

        Node current = head;
        Node pred = head;
        int count = 1;

        while(count < position && current.next != head)
        {
            pred = current;
            current = current.next;
            count++;
        }

        if(count < position)
        {
            System.out.println("Position Out of Bound");
            return;
        }

        pred.next = current.next;

        if (current == last) 
        {
            last = pred;
        }
    }
}

public class Cir_Linked_List
{ 
    public static void main(String[] args)
    {
        Scanner sc = new Scanner(System.in);

        LinkedList li = new LinkedList();

        int ch;

        do
        {
            System.out.println("Enter 1 for Add at Front");
            System.out.println("Enter 2 for Insert at End");
            System.out.println("Enter 3 for Display");
            System.out.println("Enter 4 for Delete Node from Specified Position");
            System.out.println("Enter 5 for Exit");

            System.out.print("Enter your choice: ");
            ch = sc.nextInt();


            switch(ch)
            {
                case 1:
                        System.out.println("Enter data to add at the front: ");
                        li.addFront(sc.nextInt());
                        break;
                case 2:
                        System.out.println("Enter data to insert at the end: ");
                        li.insertEnd(sc.nextInt());
                        break;
                case 3:
                        System.out.print("Display Circular Linked List-->");
                        li.display();
                        break;
                case 4:
                        System.out.println("Enter Position to delete: ");
                        int position = sc.nextInt();
                        li.deleteAtPosition(position);
                        break;
                case 5:
                        System.exit(0);
                
                default:
                        System.out.println("Plz Enter Valid Number....");
            }
        }while(ch!=5);
    }
} 

File: DS _LAB_SEM_3/Stack_Program/Cir_Linked_List.c
#include<stdio.h>

struct node{
	int no;
	struct node *next;
}*f=NULL,*r=NULL;

void insert(void);
void display(void);
void del(void);

main()
{
	int ch;
	
	do
	{
		printf("\n 0 for Exist:");
		printf("\n 1 for Insert:");
		printf("\n 2 for Display:");
		printf("\n 3 for Delete: ");
		
		printf("\n Enter choice: ");
		scanf("%d",&ch);
		
		switch(ch)
		{
			case 0 : 
					exit(1);
						
			case 1 :
					insert();
					break;
			
			case 2 :
					display();
					break;
			
			case 3 :
					del();
					break;
					
			default: 	
					printf("Enter Valid Number");
		}
	}while(ch!=0);
	
	getch();
}

void insert (void)
{
	struct node *n;
	
	n = (struct node *)malloc(sizeof(struct node));
	
	printf("\n Enter Value-> ");
	scanf("%d",&n->no);
	
	if(f==NULL && r==NULL)
	{
		f=n;
		r=n;
		r->next=f;
	}
	else
	{
		r->next=n;
		r=n;
		r->next=f;
	}
	getch();
}

void display(void)
{
	struct node *i;
	
	if(f==NULL && r==NULL)
	{
		printf("\n List Is Empty");
	}
	else
	{
		i=f;
		do
		{
			printf("\n %d",i->no);
			i=i->next;
		}while(i!=f);
	}
	getch();
}

void del(void)
{
	struct node *i;
	struct node *j;
	int num;
	int flag=0;
	
	printf("\n Enter Number to Delete: ");
	scanf("%d",&num);
	
	//1 NODE EXIST AND THEN IT DELETE
	if(f->no==num && f->next==NULL)
	{
		f=NULL;
		r=NULL;
		flag=1;
	}//1st NODE WILL BE DELETED from list
	else if(f->no==num && f->next!=NULL)
	{
		f=f->next;
		r->next=f;
		flag=1;
	}
	else
	{
		//searching
		i=f;
		while(i->no!=num && i->next!=NULL)
		{
			j=i;
			i=i->next;
		}
		
		//Delte node from list into Middle
		if(i->no==num && i!=r)
		{
			j->next=i->next;
			flag=1;
		}
		
		//Delte node from list at the end
		if(i->no==num && i==r)
		{
			r=j;
			r->next=f;
			flag=1;
		}
	}
	
	if(flag==0)
	{
		printf("\n Value Not Found");
	}
	getch();
}
